src/orange/construct/OrangeInput.hh

```diff
@@ -97,13 +102,38 @@ struct UnitInput
     explicit operator bool() const { return !volumes.empty(); }
 };
 
+//---------------------------------------------------------------------------//
+/*!
+ * Input definition for a rectangular array universe.
+ */
+struct RectArrayInput
+{
+    // Grid boundaries in x, y, and z
+    Array<std::vector<double>, 3> grid;
+
+    // Daughter loading
+    std::vector<DaughterInput> daughters;
+
+    BoundingBox bbox;  //!< Outer bounding box
+
+    // Unit metadata
+    Label label;
+
+    //! Whether the universe definition is valid
+    explicit operator bool() const { return !daughters.empty(); }
+};
+
 //---------------------------------------------------------------------------//
 /*!
  * Construction definition for a full ORANGE geometry.
  */
 struct OrangeInput
 {
     std::vector<UnitInput> units;
+    std::vector<RectArrayInput> rect_arrays;
+
+    std::vector<UniverseType> universe_types;
+    std::vector<size_type> universe_indices;
```

Comment:
Since we require C++17 now, what do you think about using a `std::vector<std::variant<UnitInput , RectArrayInput>>`? I bet this would simplify the JSON conversion *and* the input construction. I bet we could use `std::visit` and a helper lambda/functor class when looping over the input to elegantly reduce code duplication.

Reply:
Celeritas requires C++17, but does SCALE? The `OrangeInput` struct is what is used by `Celeritas_Converter` within `Geometria`, so adding a vector of `std::variant` would introduce a C++17 requirement. 

I also skeptical that this would significantly simplify things. No matter what, `OrangeParamsData` will have to store `simple_units`, and `rect_arrays` separate so these changes would have a limited scope. In addition, there isn't a lot of code duplication in the first place. The `from_json(nlohmann::json const& j, OrangeInput& value)` function would still need an if-else for each universe type. Within the `OrangeParams` constructor, the 3 `for` loops for inserting universes (simple, rect, hex) would be replaced with a single `for` loop with a switch statement. I think there is more value in having `OrangeInput` be structured similar to `OrangeParams` data than there is in introducing more complex data structures for marginal improvements.

Reply:
Yes, SCALE has required C++17 for a while. My thought was that the `OrangeInput` should be structured more along the lines of *user* input than the params input (since that could be refactored someday), but since we have to do a lot of preprocessing even before the input is created, that's not necessarily the case.

The benefit of using a `variant` is that the additional complexity of the `switch` statements could be hidden in an overload and invoked with `std::visit` , so the construction code becomes more readable.